### PR TITLE
[l2-load-balancer] Deprecating the module

### PR DESCRIPTION
**Pull request by tarabrind**
## Description
Module `l2LoadBalancer` is deprecated. Add an alert about it.

## Why do we need it, and what problem does it solve? 
It is necessary to warn and prepare for the transfer of clients to use the MetalLB module.

## What is the expected result?
- Users are warned that the module is deprecated.
- Not allowed to update cluster if the l2-load-balancer module is enabled.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: l2-load-balancer
type: feature
summary: Set the l2-load-balancer module to deprecated status.
impact: l2-load-balancer module is deprecated. To upgrade to the next DKP version, you must disable the l2-load-balancer module.
impact_level: high
```

### CODE REVIEW

File: ee/modules/381-l2-load-balancer/hooks/alert_module_deprecated_l2lb.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+	err := sdk.FromUnstructured(obj, &l2loadbalancer)
+	if err != nil {
+		return nil, err
+	}
+	return true, nil
+}
+
+func alertModuleDeprecatedL2lbExists(input *go_hook.HookInput) error {
+	if len(input.Snapshots["l2loadbalancers"]) > 0 {
```
Comment (Beastlex) [inline ee/modules/381-l2-load-balancer/hooks/alert_module_deprecated_l2lb.go:38]: use len(...) to specify in the alert message how many such objects are in the cluster

**Review by Beastlex — CHANGES_REQUESTED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Beastlex. I've started the workflow for label 'e2e/run/static'...

File: .github/dependabot.yml [hunk 2]
```diff
@@ -6,692 +6,8 @@ version: 2
 
 updates:
 # go.mod
-- package-ecosystem: "gomod"
```
Comment (AndreyPavlovFlant) [inline .github/dependabot.yml:9]: In this case, there should be no changes to this file. Something broke when running `make generate` locally. Try to revert the changes to this file and run `make generate` again.

**Review by AndreyPavlovFlant — CHANGES_REQUESTED** [review]

File: ee/modules/381-l2-load-balancer/hooks/alert_module_deprecated_l2lb.go (hunk 1 above)
Comment (tarabrind) [reply ee/modules/381-l2-load-balancer/hooks/alert_module_deprecated_l2lb.go:38]: One metric per two hooks is used, so applying len is not a good idea.

File: .github/dependabot.yml (hunk 2 above)
Comment (tarabrind) [reply .github/dependabot.yml]: Done

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @tarabrind. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @tarabrind. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: ee/modules/381-l2-load-balancer/requirements/check.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+const (
+	l2LoadBalancerModuleDeprecatedKey                = "l2LoadBalance:isEnadledModule"
+	l2LoadBalancerModuleDeprecatedKeyRequirementsKey = "l2LoadBalanceIsDeprecatedModule"
+)
+
+func init() {
+	checkRequirementIsDeprecatedModule := func(_ string, getter requirements.ValueGetter) (bool, error) {
+		hasDeprecatedModule, exists := getter.Get(l2LoadBalancerModuleDeprecatedKey)
+		if exists {
```
Comment (Beastlex) [inline ee/modules/381-l2-load-balancer/requirements/check.go:22]:
```suggestion
		if exists && hasDeprecatedModule.(bool) {
```

**Review by Beastlex — CHANGES_REQUESTED** [review]

File: ee/modules/381-l2-load-balancer/hooks/alert_module_deprecated.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	"strconv"
+
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook/metrics"
+	"github.com/flant/addon-operator/sdk"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	OnBeforeAll: &go_hook.OrderedConfig{Order: 10},
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/hooks/alert_module_deprecated.go:18]:
```suggestion
	OnStartup: &go_hook.OrderedConfig{Order: 10},
```

File: ee/modules/381-l2-load-balancer/monitoring/prometheus-rules/services.yaml [hunk 5]
```diff
@@ -16,3 +16,20 @@
...
+        tier: application
+      annotations:
+        plk_markup_format: "markdown"
+        plk_protocol_version: "1"
+        plk_create_group_if_not_exists__istio_irrelevant_external_services: ClusterHasL2LoadBalancerDeprecatedModule,tier=~tier,prometheus=deckhouse,kubernetes=~kubernetes
+        plk_grouped_by__istio_irrelevant_external_services: ClusterHasL2LoadBalancerDeprecatedModule,tier=~tier,prometheus=deckhouse,kubernetes=~kubernetes
+        summary: L2LoadBalancer module is deprecated
+        description: |
+          The L2LoadBalancer module is deprecated and will be removed in a future release. Use MetalLB module in L2 mode.
+          Now L2LoadBalancer module is enabled, there are {{$labels.l2lb_counter}} `L2LoadBalancer` in the cluster.
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/monitoring/prometheus-rules/services.yaml:35]:
```suggestion
          Now L2LoadBalancer module is enabled, also, there are {{$labels.l2lb_counter}} `L2LoadBalancer` resources in the cluster.
```

File: ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go [hunk 6]
```diff
@@ -15,6 +15,12 @@ import (
 	"github.com/flant/addon-operator/sdk"
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/deckhouse/deckhouse/go_lib/dependency/requirements"
+)
+
+const (
+	l2LoadBalancerModuleDeprecatedKey = "l2LoadBalance:isEnadledModule"
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go:23]:
```suggestion
	l2LoadBalancerModuleDeprecatedKey = "l2LoadBalancer:isModuleEnabled"
```

File: ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go [hunk 7]
```diff
@@ -136,6 +142,9 @@ func applyLoadBalancerFilter(obj *unstructured.Unstructured) (go_hook.FilterResu
 }
 
 func handleLoadBalancers(input *go_hook.HookInput) error {
+	// Set requirements values that the module is enabled
+	requirements.SaveValue(l2LoadBalancerModuleDeprecatedKey, true)
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go:146]: to ee/modules/381-l2-load-balancer/hooks/alert_module_deprecated.go hook.

File: ee/modules/381-l2-load-balancer/requirements/check.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"errors"
+
+	"github.com/deckhouse/deckhouse/go_lib/dependency/requirements"
+)
+
+const (
+	l2LoadBalancerModuleDeprecatedKey                = "l2LoadBalance:isEnadledModule"
+	l2LoadBalancerModuleDeprecatedKeyRequirementsKey = "l2LoadBalanceIsDeprecatedModule"
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/requirements/check.go:16]:
```suggestion
	l2LoadBalancerModuleDeprecatedKey                = "l2LoadBalancer:isModuleEnabled"
	l2LoadBalancerModuleDeprecatedKeyRequirementsKey = "l2LoadBalancerModuleEnabled"
```

**Review by apolovov — CHANGES_REQUESTED** [review]

**Comment by tarabrind** [comment]
Module is enabled.
```
[dev-rnd] root@virtlab-ap-0:~# kubectl get clusteralerts.deckhouse.io
NAME               ALERT                                      SEVERITY   AGE     LAST RECEIVED   STATUS
9be9827254048ad6   L2LoadBalancerModuleDeprecated             3          58s     28s             firing

[dev-rnd] root@virtlab-ap-0:~# kubectl get deckhousereleases.deckhouse.io
NAME       PHASE     TRANSITIONTIME   MESSAGE
v1.101.9   Pending   3s               "l2LoadBalancerModuleEnabled" requirement for DeckhouseRelease "1.101.9" not met: the L2LoadBalancer module is deprecated and will be removed in a future release. Use MetalLB module in L2 mode
```

Module is disabled.

```
[dev-rnd] root@virtlab-ap-0:~# kubectl get deckhousereleases.deckhouse.io
NAME       PHASE      TRANSITIONTIME   MESSAGE
v1.101.9   Deployed   6s
```

**Review by apolovov — APPROVED** [review]

**Review by AndreyPavlovFlant — APPROVED** [review]

**Review by Beastlex — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Cherry pick PR [9814](https://github.com/deckhouse/deckhouse/pull/9814) to the branch [release-1.64](https://github.com/deckhouse/deckhouse/tree/release-1.64) successful!
